### PR TITLE
feat(Run): Allow teachers to archive runs

### LINKED ISSUE #1012: feat(Run): Allow teachers to archive units/runs
## Description
Provide a mechanism for teachers to archive (and restore) runs and units that they no longer want to see in their Class Schedule and My Units library.

### PR DESCRIPTION
**Pull request by geoffreykwan**
## Changes

- Teachers can now archive runs
- Removed filter by period select drop down in Class Schedule view
- There is now an active runs view and an archived runs view

## Test

- Test with https://github.com/WISE-Community/WISE-API/pull/225

- Archive/unarchive a run using the run item menu list (the 3 dots menu on a run)
- Archive/unarchive a run using the run item checkboxes and then the "Archive Selected"/"Unarchive Selected" button
- Archive/unarchive a run that was shared with you
- Make sure the select all runs checkbox works properly
- Make sure the select runs drop down works properly for (all, none, running, completed)
- Make sure switching from the active to archived view works properly

Closes #1012

### CODE REVIEW

File: src/app/teacher/teacher-run-list/teacher-run-list.component.html [hunk 1]
```diff
@@ -58,23 +58,59 @@
...
+    <button
+      mat-raised-button
+      color="accent"
+      *ngIf="showArchived && numSelectedRuns > 0"
+      (click)="unarchiveSelectedRuns()"
+      i18n
+    >
+      Unarchive Selected
+    </button>
+  </div>
```
Comment (hirokiterashima) [inline src/app/teacher/teacher-run-list/teacher-run-list.component.html]: TeacherRunListComponent is getting big. Can we make this block into a new component?

File: src/app/teacher/run-menu/run-menu.component.ts [hunk 2]
```diff
@@ -86,4 +90,32 @@ export class RunMenuComponent implements OnInit {
...
+      next: (response: ArchiveProjectResponse) => {
+        run.archived = response.archived;
+        this.runArchiveStatusChangedEvent.emit(run);
+        this.snackBar.open($localize`Successfully Unarchived Run`);
+      },
+      error: () => {
+        this.snackBar.open($localize`Error Unarchiving Run`);
+      }
+    });
+  }
```
Comment (hirokiterashima) [inline src/app/teacher/run-menu/run-menu.component.ts]: Duplication. Can you try refactoring it?

File: src/app/services/archive-project.service.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  unarchiveProjects(projects: Project[]): Observable<any> {
+    let params = new HttpParams();
+    for (const project of projects) {
+      params = params.append('projectIds', project.id);
+    }
+    return this.http.delete<any>(`/api/projects/archived`, {
+      params: params
+    });
+  }
```
Comment (hirokiterashima) [inline src/app/services/archive-project.service.ts]: Instead of ```Observable<any>```, is it ```Observable<ArchiveProjectResponse[]>```?

**Review by hirokiterashima — COMMENTED** [review]
I added some comments. Also, should we use "Archive/Unarchive Unit" instead of "Archive/Unarchive Run"? We want teachers to think in "unit", not "run"?

**Review by breity — CHANGES_REQUESTED** [review]
Looking great. I updated styles and layout, as well as some text, accessibility elements, etc.
@geoffreykwan There are two failing tests that I'm having trouble fixing. Maybe we could take a look.

I have 1 suggestion: We should add a confirmation dialog when user chooses to archive or restore multiple runs at once. @hirokiterashima and @geoffreykwan, what do you think?

These are the issues I found that we discussed:
1. Selecting all the runs using the top checkbox doesn't select shared runs.
2. All the user's runs are shown now when loading the run list. We used to only show the latest 10 and let users view the rest with the 'Show All' button.

**Comment by hirokiterashima** [comment]
New style changes look good. 

I wonder if asking users to confirm each time is a bit disruptive, especially since it's easy to undo their changes in the archive/unarchive views if they make a mistake by accident. How about adding an 'undo' button in the toast message, like in Gmail?

**Comment by hirokiterashima** [comment]
The new undo feature works great.

Should we hide or disable the SelectRunsControls if there are no runs to select? It seems a bit strange to have it there.
Screenshot 2023-08-28 at 11 20 42 AM

**Comment by breity** [comment]
> Should we hide or disable the SelectRunsControls if there are no runs to select? It seems a bit strange to have it there. 

Ah, yep. Good catch. We should hide them.
EDIT: Actually, none of that should be visible when there aren't any active runs. User should see the notice with text "Hey there! Looks like you haven't run any WISE units in your classes yet..." It's this code in TeacherRunListComponent:
```
<div class="notice" *ngIf="!runs.length && loaded">
  <p i18n>Hey there! Looks like you haven't run any WISE units in your classes yet.</p>
  <p i18n>Select "Unit Library" to find titles to use with your students.</p>
</div>
<div class="content-block" *ngIf="runs.length">
...
```

We should update the *ngIf to look for any active runs, rather than any runs for the user. We should also add a different notice when viewing the archive and the user has no archived runs (i.e. "You don't have any archived classroom units.").

File: src/app/teacher/run-menu/run-menu.component.ts [hunk 4]
```diff
@@ -8,23 +8,28 @@ import { ConfigService } from '../../services/config.service';
...
   styleUrls: ['./run-menu.component.scss']
 })
 export class RunMenuComponent implements OnInit {
+  private editLink: string = '';
+  protected reportProblemLink: string = '';
   @Input() run: TeacherRun;
-
-  editLink: string = '';
-  reportProblemLink: string = '';
+  @Output() runArchiveStatusChangedEvent: EventEmitter<any> = new EventEmitter<any>();
```
Comment (hirokiterashima) [inline src/app/teacher/run-menu/run-menu.component.ts]: Looks like we don't do anything with the emitted value, so make it void?
```suggestion
  @Output() runArchiveStatusChangedEvent: EventEmitter<void> = new EventEmitter<void>();
```

File: src/app/teacher/run-menu/run-menu.component.ts [hunk 5]
```diff
@@ -86,4 +83,57 @@ export class RunMenuComponent implements OnInit {
...
+    );
+  }
+
+  private showErrorMessage(archive: boolean): void {
+    this.snackBar.open($localize`Error ${archive ? 'archiving' : 'unarchiving'} unit.`);
+  }
+
+  private updateArchivedStatus(run: TeacherRun, archived: boolean): void {
+    run.archived = archived;
+    this.runArchiveStatusChangedEvent.emit(run);
```
Comment (hirokiterashima) [inline src/app/teacher/run-menu/run-menu.component.ts]: Looks like we don't do anything with the emitted value, so make it void?
```suggestion
    this.runArchiveStatusChangedEvent.emit();
```

File: src/app/teacher/teacher-run-list-item/teacher-run-list-item.component.ts [hunk 6]
```diff
@@ -15,13 +15,14 @@ import { ShareRunCodeDialogComponent } from '../share-run-code-dialog/share-run-
...
+  protected periodsTooltipText: string;
   @Input() run: TeacherRun = new TeacherRun();
-
-  manageStudentsLink: string = '';
-  periodsTooltipText: string;
-  thumbStyle: SafeStyle;
-  animateDuration: string = '0s';
-  animateDelay: string = '0s';
+  @Output() runArchiveStatusChangedEvent: EventEmitter<any> = new EventEmitter<any>();
+  @Output() runSelectedStatusChangedEvent: EventEmitter<any> = new EventEmitter<any>();
```
Comment (hirokiterashima) [inline src/app/teacher/teacher-run-list-item/teacher-run-list-item.component.ts]: Looks like we don't do anything with the emitted value, so make it void?
```suggestion
  @Output() runArchiveStatusChangedEvent: EventEmitter<void> = new EventEmitter<void>();
  @Output() runSelectedStatusChangedEvent: EventEmitter<void> = new EventEmitter<void>();
```

**Review by hirokiterashima — CHANGES_REQUESTED** [review]
It's getting there. As discussed this morning, here are some more changes to make:

1. In both active and archive views, add “scheduled” select-option, and show the option order as follows: All -> None -> Completed -> Running -> Scheduled

2. Change message when there are no active runs to “Hey there! Looks like you don't have any active classroom units.\n\nBrowse the "Unit Library" to find titles to use with your students.” Also remove the count message.

3. Change message when there are no archived runs to “You don’t have any archived classroom units.” Also remove the count message.

4. See minor code comments inline

File: src/app/teacher/teacher-run-list/teacher-run-list.component.spec.ts [hunk 7]
```diff
@@ -1,130 +1,435 @@
...
+}
+
+function selectScheduledOptionChosen(): void {
+  describe('when scheduled is chosen', () => {
+    it('it should select scheduled runs', async () => {
+      setRun3Scheduled();
+      await clickSelectRunsMenuButtonAndExpectSelected('Scheduled', [true, false, false]);
+    });
+  });
+}
```
Comment (hirokiterashima) [inline src/app/teacher/teacher-run-list/teacher-run-list.component.spec.ts]: This block introduces duplication. How about writing it this way? Also, let's read and discuss [Jasmine style guide](https://github.com/CareMessagePlatform/jasmine-styleguide) to standardize how we structure/write unit tests.
```suggestion
function selectRunsOptionChosen(): void {
  describe('selectRunsOptionChosen()', () => {
    beforeEach(() => {
      setRun2Completed();
    });
    selectAllOptionChosen();
    selectNoneOptionChosen();
    selectCompletedOptionChosen();
    selectRunningOptionChosen();
    selectScheduledOptionChosen();
  });
}

function selectAllOptionChosen(): void {
  it('should select all runs', async () => {
    await clickSelectRunsMenuButtonAndExpectSelected('All', [true, true, true]);
  });
}

function selectNoneOptionChosen(): void {
  it('should select no runs', async () => {
    await clickSelectRunsMenuButtonAndExpectSelected('None', [false, false, false]);
  });
}

function selectCompletedOptionChosen(): void {
  it('should select completed runs', async () => {
    await clickSelectRunsMenuButtonAndExpectSelected('Completed', [false, true, false]);
  });
}

function selectRunningOptionChosen(): void {
  it('should select running runs', async () => {
    await clickSelectRunsMenuButtonAndExpectSelected('Running', [true, false, true]);
  });
}

function selectScheduledOptionChosen(): void {
  it('it should select scheduled runs', async () => {
    setRun3Scheduled();
    await clickSelectRunsMenuButtonAndExpectSelected('Scheduled', [true, false, false]);
  });
}
```

File: src/app/teacher/teacher-run-list/teacher-run-list.component.ts [hunk 8]
```diff
@@ -212,8 +200,51 @@ export class TeacherRunListComponent implements OnInit {
...
+        case 'running':
+          run.selected = run.isActive(this.configService.getCurrentServerTime());
+          break;
+        case 'scheduled':
+          run.selected = run.isScheduled(this.configService.getCurrentServerTime());
+          break;
+      }
+    });
+    this.runSelectedStatusChanged();
+  }
```
Comment (hirokiterashima) [inline src/app/teacher/teacher-run-list/teacher-run-list.component.ts]: Put the logic in TeacherRun?
```suggestion
  protected selectRunsOptionChosen(option: SelectRunsOption): void {
    const now = this.configService.getCurrentServerTime();
    this.filteredRuns.forEach((run: TeacherRun) => run.updateSelected(option, now));
    this.runSelectedStatusChanged();
  }
```

File: src/app/teacher/teacher-run.ts [hunk 9]
```diff
@@ -1,8 +1,10 @@
 import { Run } from '../domain/run';
 
 export class TeacherRun extends Run {
```
Comment (hirokiterashima) [inline src/app/teacher/teacher-run.ts:4]: Move code for setting teacherRun.selected into this class?
```
  updateSelected(selectRunsOption: SelectRunsOption, currentTime: number): void {
    switch (selectRunsOption) {
      case 'ALL':
        this.selected = true;
        break;
      case 'NONE':
        this.selected = false;
        break;
      case 'COMPLETED':
        this.selected = this.isCompleted(currentTime);
        break;
      case 'RUNNING':
        this.selected = this.isActive(currentTime);
        break;
      case 'SCHEDULED':
        this.selected = this.isScheduled(currentTime);
        break;
    }
  }
```

File: src/app/teacher/select-runs-controls/select-runs-controls.component.ts [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+  providers: [{ provide: MAT_CHECKBOX_DEFAULT_OPTIONS, useValue: { clickAction: 'noop' } }]
+})
+export class SelectRunsControlsComponent {
+  @Output() archiveActionEvent = new EventEmitter<void>();
+  protected numSelectedRuns: number = 0;
+  @Input() runChangedEventEmitter: EventEmitter<void> = new EventEmitter<void>();
+  @Input() runs: TeacherRun[] = [];
+  protected selectedAllRuns: boolean = false;
+  protected selectedSomeRuns: boolean = false;
+  @Output() selectRunsOptionChosenEvent = new EventEmitter<string>();
```
Comment (hirokiterashima) [inline src/app/teacher/select-runs-controls/select-runs-controls.component.ts]: Emit SelectRunsOptions enum class?
```suggestion
  @Output() selectRunsOptionChosenEvent = new EventEmitter<SelectRunsOption>();
```

SelectRunsOptions enum class definition:
```
export enum SelectRunsOption {
  All = 'ALL',
  None = 'NONE',
  Completed = 'COMPLETED',
  Running = 'RUNNING',
  Scheduled = 'SCHEDULED'
}
```

File: src/app/teacher/select-runs-controls/select-runs-controls.component.ts [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+  protected selectAllRunsCheckboxClicked(): void {
+    this.selectRunsOptionChosenEvent.emit(
+      this.selectedAllRuns || this.selectedSomeRuns ? 'none' : 'all'
+    );
+  }
+
+  protected selectRunsOptionChosen(value: string): void {
+    this.selectRunsOptionChosenEvent.emit(value);
+  }
```
Comment (hirokiterashima) [inline src/app/teacher/select-runs-controls/select-runs-controls.component.ts]: Emit SelectRunsOption enum value?
```suggestion
  protected selectAllRunsCheckboxClicked(): void {
    this.selectRunsOptionChosenEvent.emit(
      this.selectedAllRuns || this.selectedSomeRuns ? SelectRunsOption.None : SelectRunsOption.All
    );
  }

  protected selectRunsOptionChosen(value: string): void {
    this.selectRunsOptionChosenEvent.emit(value as SelectRunsOption);
  }
```

File: src/app/teacher/select-runs-controls/select-runs-controls.component.html [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+      i18n-aria-label
+    >
+      <mat-icon>arrow_drop_down</mat-icon>
+    </button>
+    <mat-menu #selectRunsMenu="matMenu">
+      <button mat-menu-item (click)="selectRunsOptionChosen('all')" i18n>All</button>
+      <button mat-menu-item (click)="selectRunsOptionChosen('none')" i18n>None</button>
+      <button mat-menu-item (click)="selectRunsOptionChosen('completed')" i18n>Completed</button>
+      <button mat-menu-item (click)="selectRunsOptionChosen('running')" i18n>Running</button>
+      <button mat-menu-item (click)="selectRunsOptionChosen('scheduled')" i18n>Scheduled</button>
```
Comment (hirokiterashima) [inline src/app/teacher/select-runs-controls/select-runs-controls.component.html]: Use enum values?
```suggestion
      <button mat-menu-item (click)="selectRunsOptionChosen('ALL')" i18n>All</button>
      <button mat-menu-item (click)="selectRunsOptionChosen('NONE')" i18n>None</button>
      <button mat-menu-item (click)="selectRunsOptionChosen('COMPLETED')" i18n>Completed</button>
      <button mat-menu-item (click)="selectRunsOptionChosen('RUNNING')" i18n>Running</button>
      <button mat-menu-item (click)="selectRunsOptionChosen('SCHEDULED')" i18n>Scheduled</button>
```

**Review by hirokiterashima — CHANGES_REQUESTED** [review]
I added some code suggestions to prevent adding new code duplication issues in the TeacherRunListComponent spec file. I referenced [Jasmine style guide](https://github.com/CareMessagePlatform/jasmine-styleguide).

I also think using a new SelectRunsOption enum to represent available options will make the code clearer, so I added some suggestions inline for that as well. Lmk what you think or have any questions.

**Review by hirokiterashima — APPROVED** [review]
LGTM

**Review by breity — APPROVED** [review]
Looks good!

**Pull request merged**